informer-gen: correctly generate group packages when groups contain hyphens

munnerz opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

informer-gen currently can generate invalid package names containing hyphens if a groupName whose first segment contains a hyphen is used, e.g. `cert-manager.io`.

This is because currently informer-gen uses the first segment of the group name as specified with the `groupName` annotation as the package name.
This patch updates informer-gen to instead use the computed PackageName based on the `--input-dirs` flag, similar to client-gen et al.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/cc @smarterclayton @sttts

munnerz committed:
informer-gen: use correct PackageName computed from input parameters

munnerz commented:
/sig api-machinery
/priority important-soon

sttts commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/packages.go:
```diff
@@ -284,6 +284,10 @@ func factoryInterfacePackage(basePackage string, boilerplate []byte, clientSetPa
+	// It is never valid to have a hyphen in package names, so remove all
+	// hyphens from package names to make the default path for users with
+	// group names containing hyphens better.
+	groupPkgName = strings.ReplaceAll(groupPkgName, "-", "")
```
is this the same client-gen does?

munnerz replied:
client-gen relies on the `--input` parameter to define the group name, which is then used as the package name. The code for this *is* reusable (see here: https://github.com/kubernetes/kubernetes/blob/7e806a2dd6efb156858706aca78f8a9850315f7b/staging/src/k8s.io/code-generator/cmd/client-gen/args/gvpackages.go)

*HOWEVER*: informer-gen doesn't accept a --input argument, so this would require quite significantly more changes to make it work/possible (incl. changing the CLI for informer-gen).

I'm not sure how we want to proceed with this 😬

munnerz committed:
Add _example & test for a hyphenated API group

Signed-off-by: James Munnelly <james@munnelly.eu>

munnerz commented:
I've updated the PR to no longer just do a blind string replacement and instead use the computed PackageName based on the input arguments passed to informer-gen. This brings it in-line with what other generators do.

I've also added a new _example for this to demonstrate it working 😄

munnerz committed:
Regenerate _example informers packages

Signed-off-by: James Munnelly <james@munnelly.eu>

munnerz commented:
I've added https://github.com/kubernetes/kubernetes/pull/80440/commits/cab04acf37a9f671a7109497082fe137badf793b which fixes the verify test failure.

It also changes the package name for the informers package for the `example2` group from `example` to `example2`, which IMO is the 'correct' name (given that the directory it is contained in is already named `example2`). This would not have been required *iff* the group name was previously correctly set in the example2 package, however it currently specifies `// +groupName=example.test.apiserver.code-generator.k8s.io` meaning the old informers code would set the package name to `example` instead of `example2` (despite the directory being named `example2`).

I've therefore also pushed up [067cbe9e39](https://github.com/kubernetes/kubernetes/pull/80440/commits/067cbe9e39d3f6f6771da6003aca0fa0d5699242) which updates the groupName annotations to be example2 instead.

munnerz commented:
/retest

munnerz commented:
I'm assuming this test is failing due to a flake.. I'd be very surprised if this change *only* caused the gpu device plugin tests to fail 😅

munnerz commented:
/retest

munnerz commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/packages.go:
```diff
@@ -283,10 +283,9 @@ func factoryInterfacePackage(basePackage string, boilerplate []byte, clientSetPa
 
 	return &generator.DefaultPackage{
-		PackageName: groupPkgName,
+		PackageName: groupVersions.PackageName,
```
This one line is the 'meat' of the PR 😄

roycaihw commented:
/cc @yliaog

yliaog commented on staging/src/k8s.io/code-generator/_examples/HyphenGroup/apis/example/v1/doc.go:
```diff
@@ -0,0 +1,21 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
s/2016/2019/

yliaog commented on staging/src/k8s.io/code-generator/_examples/HyphenGroup/apis/example/v1/register.go:
```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
s/2015/2019/

yliaog commented on staging/src/k8s.io/code-generator/_examples/HyphenGroup/apis/example/v1/types.go:
```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
s/2015/2019/

munnerz committed:
Update copyright years

Signed-off-by: James Munnelly <james@munnelly.eu>

munnerz commented:
Updated copyright years

yliaog commented:
/lgtm

sttts commented on staging/src/k8s.io/code-generator/_examples/apiserver/apis/example2/doc.go:
```diff
@@ -15,7 +15,7 @@ limitations under the License.
 
 // +k8s:deepcopy-gen=package
-// +groupName=example.test.apiserver.code-generator.k8s.io
+// +groupName=example2.test.apiserver.code-generator.k8s.io
```
this was intentionally called `example`

munnerz commented:
@sttts I've reverted that change.

This patch means the name of the package that informer-gen generates for the `example.test.apiserver.code-generator.k8s.io` group will change to `example2` (including the directory name), however this *is* consistent with `client-gen` and `lister-gen` which also generate packages named `example2` 😄

sttts commented:
/lgtm
/approve

munnerz commented:
/retest
